### ISSUE TITLE
adding sample weights

### ISSUE THREAD

**Issue by spedygiorgio** [soda-inria/tabicl#115]
# Feature Request: Support for `sample_weight` and Custom Losses in `TabICLRegressor`

Hi,

first of all, thank you for the great work on TabICL.  
I am experimenting with `TabICLRegressor` on actuarial and insurance pricing datasets, and I would like to ask whether support for `sample_weight` and customizable loss functions could be added.

## Motivation

In insurance modeling, observations often have different statistical credibility or exposure.

For example, when modeling claim frequency:

```math
\text{frequency} = \frac{\text{number of claims}}{\text{exposure}}
```

it is standard practice to weight observations by exposure:

```python
model.fit(X, y, sample_weight=exposure)
```

This is equivalent to the classical actuarial Poisson formulation.

Currently, it seems that `TabICLRegressor.fit()` may not support `sample_weight`, and there does not appear to be a way to specify alternative regression losses or objectives.

### 1. Support for `sample_weight`

Scikit-learn-style API support such as:

```python
model.fit(X, y, sample_weight=weights)
```

would be extremely useful for:

- actuarial modeling
- imbalanced regression
- heteroskedastic data
- reliability-weighted observations

---

### 2. Customizable Regression Losses

It would also be very valuable to support alternative objectives/losses such as:

- Poisson deviance
- Gamma deviance
- Tweedie loss
- MAE / Huber loss
- user-defined custom losses

For example:

```python
TabICLRegressor(loss="poisson")
```

or:

```python
TabICLRegressor(loss="tweedie")
```

This would significantly broaden the applicability of TabICL to:

- insurance pricing
- risk modeling
- count data
- skewed positive targets

## Additional Note

Foundation tabular models are very promising for structured data, and adding weighted fitting plus generalized losses would make TabICL much more competitive with models such as:

- CatBoost
- LightGBM
- XGBoost

especially in actuarial and industrial use cases.

Thanks again for the project and the excelle

**Comment by dholzmueller** [comment]
Hi Giorgio,
your post is timely as I just yesterday read this blog post: https://oleksandrruppelt.substack.com/p/insurance-pricing-with-tabpfn-30

Regarding sample weights: This is not a trivial thing to do in the PFN/ICL paradigm. It depends on the meaning of the sample weights. In your case I assume that the reason for the sample weights is that observations with lower sample weight have higher noise variance and you don't want them to have too strong of an influence on the learning process?
The blog post used exposure as a covariate instead. Doing that sounds like a decent workaround to me until we figure out how to incorporate sample weights.

Regarding alternative losses: In principle, TabICL should learn to approximate the predictive posterior distribution. The optimizers of these losses could be derived from that. For example, for Tweedie it would be the mean of the posterior distribution and for MAE it would be the median. The blogpost says that it doesn't work well for Tweedie and I don't fully understand the reasons for that. Do you observe a similar thing?
Of course, introducing distributions as observed in insurance pricing into the prior might help to improve the predictions on such datasets, but I don't see why different pre-training losses should make a large difference.

**Comment by mayer79** [comment]
Additional to using log exposure weights as features, one can sample a subset of the training data based on exposure weights.

**Comment by spedygiorgio** [comment]
@dholzmueller thank you for your response and for the detailed explanation.

Actually, sample weights are extremely important in actuarial modeling.
For example, when modeling claim frequency, two observations may have the same target value but very different statistical credibility.

For instance:

* (10/100 = 0.10)
* (100/1000 = 0.10)

Both observations have the same empirical frequency, but the second one is based on a much larger exposure and therefore has substantially lower variance. In classical actuarial modeling, the first observation should therefore have much less influence during training.

This is why exposure is usually introduced as a sample weight rather than only as a covariate.

As far as I know, both TensorFlow and PyTorch datasets/loaders already support observation-level weights (including zero weights). Would leveraging weighted losses during training be a viable approach within the PFN/ICL paradigm, or do you think this would fundamentally conflict with the current pretraining setup?

**Comment by dholzmueller** [comment]
> Would leveraging weighted losses during training be a viable approach within the PFN/ICL paradigm

There is no training, only pretraining. In pretraining, I guess the naive approach would be to introduce random sample weights, feed them into the model, and weigh the losses with them. However, I think this would leave the Bayes-optimal solution the same (namely, Bayesian posterior predictive distributions that ignore the sample weights). There are some better ideas but I'll leave them for a future paper or so :)

In your case, would you make the sample weights proportional to the exposure? Because if you have a very large exposure, you have lower variance but there might still be other sources of variance / aleatoric uncertainty, like unobserved covariates, so maybe the sample weight shouldn't go to infinity with infinite exposure.

**Comment by yushen611** [comment]
> Hi Giorgio, your post is timely as I just yesterday read this blog post: https://oleksandrruppelt.substack.com/p/insurance-pricing-with-tabpfn-30
> 
> Regarding sample weights: This is not a trivial thing to do in the PFN/ICL paradigm. It depends on the meaning of the sample weights. In your case I assume that the reason for the sample weights is that observations with lower sample weight have higher noise variance and you don't want them to have too strong of an influence on the learning process? The blog post used exposure as a covariate instead. Doing that sounds like a decent workaround to me until we figure out how to incorporate sample weights.
> 
> Regarding alternative losses: In principle, TabICL should learn to approximate the predictive posterior distribution. The optimizers of these losses could be derived from that. For example, for Tweedie it would be the mean of the posterior distribution and for MAE it would be the median. The blogpost says that it doesn't work well for Tweedie and I don't fully understand the reasons for that. Do you observe a similar thing? Of course, introducing distributions as observed in insurance pricing into the prior might help to improve the predictions on such datasets, but I don't see why different pre-training losses should make a large difference.

Hi @dholzmueller,

Thanks for the insightful discussion. I'm facing a very similar issue when applying TabICL to insurance risk modeling.

My target variable follows a **zero-inflated lognormal distribution** (common for claim costs). However, TabICL's performance is significantly worse than a standard LightGBM with Tweedie loss – which is a strong baseline in actuarial practice.

I suspect two possible reasons:

1. TabICL's pretraining prior may not have seen zero-inflated lognormal distributions.

2. Pretraining might assume a Gaussian likelihood (MSE loss), which is suboptimal for such skewed and zero-inflated targets.

Could you share your thoughts on whether TabICL can handle such distributions out-of-the-box? Or would it require finetuning ?

Thanks a lot!

**Comment by dholzmueller** [comment]
Hi, indeed TabICL has not seen zero-inflated lognormal distributions specifically (though I don't know how close the prior can get to generating something like it). However, it does not assume a Gaussian likelihood (the MSE loss doesn't assume that, but it is trained with pinball loss anyway).
Apparently according to your experiments TabICL cannot handle it out-of-the-box, I don't have experience with it. The blog post I linked above constructs some multi-stage procedure that apparently works much better (they used TabPFN).

**Comment by spedygiorgio** [comment]
@dholzmueller prior weights may be useful when fine tuning the model on a new dataset...

**Comment by dhrou** [comment]
Seeing this issue, I'd like to chip in by writing that in particle physics application, we're using sample weights all the time.

We've done a  study on an unweighted dataset,  comparing TabICL (and TabPFN) to XGBoost (which is a standard in particle physics) in the regime of small (for particle physics) training dataset (30 features x <=10k examples). This is documented in a recent [talk at CHEP 2026, Bangkok](https://indico.cern.ch/event/1471803/contributions/6968240/attachments/3283330/5868057/Tabular%20Foundation%20Models%20for%20particle%20physics%20with%20the%20FAIR%20Universe%20HiggsML%20Dataset.pptx) (proceedings to be released end of the summer). We've been quite impressed by TFM's performance in this regime, which is untypical in particle physics, but still sufficiently frequent to raise a lot of interest.

But to "sell" this to colleagues, support for sample weight is a big hurdle. Yes, there are techniques to "deweight" dataset, with resampling for  example, but this comes with additional variance in the training dataset, and loss in performance (it is easy to do such comparison with xgboost, which supports sample weight).
What we call a weighted dataset, it is a collection {w_i, x_i} so that for any volume V, when summing on examples belonging to V,  E(Sum_V w_i) = E(Sum_V 1) (of the unweighted dataset), where the expectation runs on i.i.d datasets.

It is too long to explain why we have weighted datasets in particle physics, but it has to do to optimisation of the use of resources when doing quantum physics computation, it would be much much more expensive to do without them. In fact we also have dataset with negative weights ! But no off-the-shelf classifier can handle them, not even xgboost.

**Comment by dholzmueller** [comment]
@dhrou Thanks for weighing in! In your explanation, are both expectations over the same distribution? Is the weighting related to different variances on different samples, or some kind of biased sampling?

**Comment by dhrou** [comment]
The expectation is on the sum of weight in a given volume, which should be the same as the number of examples in the volume for the unweighted dataset. I believe this is indeed some kind of biased sampling.

A very simplified test case, for a classifier : we have three categories, a signal one S and two background one B1 and B2. The training dataset is 33% S, 33% B1 33% B2. We want to train to classify S versus B=B1+B2. but we know that in reality B2 is 10 times less likely than B1. So what we usually do is to assign weight 0.1 to B2 and 1 to B1 (and then equalise Sum_S w = Sum_B w before training). ( In this specific case, the right way is probably to do 3-categories classification, but in general we cannot do this because we have a continuous weight distribution)
We could prune B2 examples to keep 1/10 of them, before training, but this degrades the performance, as mentioned above.

**Comment by dholzmueller** [comment]
I see. So for the classifier, you would have label shift - P(y) is modified, but P(x|y) stays the same. In that I guess you'd want to use a post-hoc correction that adjusts the probabilities (if the true and modified P(y) are known, at least). Is that analogous to your situation, or do you have e.g. a covariate shift where P(x) changes but P(y|x) stays the same?
